benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/running/RunningStatesHandler.java

```diff
@@ -0,0 +1,370 @@
+    this.testModelDAO = BenchFlowTestManagerApplication.getTestModelDAO();
+
+  }
+
+  public void determineExplorationStrategy(String testID) {
+
+    logger.info("determineExplorationStrategy with testID: " + testID);
+
+    DetermineExplorationStrategyTask task = new DetermineExplorationStrategyTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future<?> future = taskExecutorService.submit(task);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    waitForRunningTaskToComplete(testID, future, ADD_STORED_KNOWLEDGE);
+  }
+
+  public void addStoredKnowledge(String testID) {
+
+    logger.info("addStoredKnowledge with testID: " + testID);
+
+    AddStoredKnowledgeTask task = new AddStoredKnowledgeTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future<?> future = taskExecutorService.submit(task);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    try {
+
+      // wait for task to complete
+      future.get();
+
+      if (testTaskScheduler.isTerminated(testID)) {
+        // if test has been terminated we stop here
+        return;
+      }
+
+      boolean hasRegressionModel = explorationModelDAO.hasRegressionModel(testID);
+
+      TestRunningState nextState =
+          hasRegressionModel ? DETERMINE_EXECUTE_VALIDATION_SET : DETERMINE_EXECUTE_EXPERIMENTS;
+
+      testModelDAO.setTestRunningState(testID, nextState);
+
+      testTaskScheduler.handleTestState(testID);
+
+    } catch (InterruptedException | ExecutionException e) {
+      // TODO - handle properly
+      e.printStackTrace();
+    } catch (BenchFlowTestIDDoesNotExistException e) {
+      // should not happen since it was added earlier
+      logger.error("test ID does not exist - should not happen");
+    }
+
+
+  }
+
+  public void determineExecuteInitialValidationSet(String testID) {
+
+    logger.info("determineExecuteInitialValidationSet with testID: " + testID);
+
+    DetermineExecuteInitialValidationSetTask initialValidationSetTask =
+        new DetermineExecuteInitialValidationSetTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future future = taskExecutorService.submit(initialValidationSetTask);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    waitForRunningTaskToComplete(testID, future, DETERMINE_EXECUTE_EXPERIMENTS);
+
+  }
+
+  public void determineAndExecuteExperiments(String testID) {
+
+    logger.info("determineAndExecuteExperiments with testID: " + testID);
+
+    DetermineExecuteExperimentsTask task = new DetermineExecuteExperimentsTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future<?> future = taskExecutorService.submit(task);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    // we don't wait for the task to complete since the experiment-manager
```

Comment:
Probably at this stage, the waiting for this state should be exactly to wait for the wanted data from the experiment manager, that should be the details for each of the trials that actually gets executed up to the point in which the experiment manager declares the experiment as completed in any of the possible states.